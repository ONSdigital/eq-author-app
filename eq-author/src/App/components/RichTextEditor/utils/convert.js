import { convertToHTML, convertFromHTML } from "draft-convert";
import { EditorState } from "draft-js";

export const toHTML = entityMap => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return editorState => convert(editorState.getCurrentContent());
};

export const fromHTML = htmlToEntity => {
  const convert = convertFromHTML({ htmlToEntity });

  return (html, decorator) =>
    EditorState.createWithContent(convert(html), decorator);
};
