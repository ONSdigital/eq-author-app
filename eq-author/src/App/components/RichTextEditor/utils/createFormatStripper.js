import { mapKeys } from "lodash";
import { filterEditorState } from "draftjs-filters";
import { filterConfig } from "App/components/RichTextEditor/entities/PipedValue";

const mapper = {
  bold: { format: "BOLD", type: "styles" },
  emphasis: { format: "ITALIC", type: "styles" },
  list: { format: "unordered-list-item", type: "blocks" },
  heading: { format: "header-two", type: "blocks" }
};

export default function createFormatStripper(controls) {
  const filterConfiguration = {
    blocks: [],
    styles: [],
    entities: [filterConfig],
    maxNesting: 0,
    whitespacedCharacters: []
  };

  mapKeys(controls, (value, key) => {
    if (mapper[key] && controls[key]) {
      filterConfiguration[mapper[key].type].push(mapper[key].format);
    }
  });

  return editorState => filterEditorState(filterConfiguration, editorState);
}
