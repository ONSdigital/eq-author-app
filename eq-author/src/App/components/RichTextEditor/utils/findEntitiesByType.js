export default entityType => {
  return function findEntitiesByType(contentBlock, callback, contentState) {
    contentBlock.findEntityRanges(character => {
      const entityKey = character.getEntity();
      return (
        character.getEntity() !== null &&
        contentState.getEntity(entityKey).getType() === entityType
      );
    }, callback);
  };
};
