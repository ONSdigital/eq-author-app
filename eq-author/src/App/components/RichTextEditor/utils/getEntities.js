function getEntities(contentState, entityType) {
  const entities = [];

  contentState.getBlocksAsArray().forEach(block => {
    let selectedEntity = null;

    block.findEntityRanges(
      character => {
        const entityKey = character.getEntity();

        if (entityKey === null) {
          return false;
        }

        const entity = contentState.getEntity(entityKey);

        if (entityType && entity.getType() !== entityType) {
          return false;
        }

        selectedEntity = {
          entityKey,
          entity,
          blockKey: block.getKey()
        };

        return true;
      },
      (start, end) => {
        entities.push({ ...selectedEntity, start, end });
      }
    );
  });

  return entities;
}

export default getEntities;
