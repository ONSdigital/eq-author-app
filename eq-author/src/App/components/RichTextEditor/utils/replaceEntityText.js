import { SelectionState, Modifier } from "draft-js";

const getEntityRange = (block, entityKey) => {
  let range;

  // get the range for the entity key
  block.findEntityRanges(
    char => char.getEntity() === entityKey,
    (start, end) => {
      range = { start, end };
    }
  );

  return SelectionState.createEmpty(block.key).merge({
    anchorKey: block.key,
    anchorOffset: range.start,
    focusKey: block.key,
    focusOffset: range.end
  });
};

const replaceText = (contentState, entityKey, blockKey, text, inlineStyles) => {
  const block = contentState.getBlockForKey(blockKey);
  const entitySelection = getEntityRange(block, entityKey);

  // first replace text
  const updatedContentState = Modifier.replaceText(
    contentState,
    entitySelection,
    text,
    inlineStyles,
    entityKey
  );

  // then create new selection, which covers the updated entity
  const updatedEntitySelection = entitySelection.merge({
    focusOffset: entitySelection.getStartOffset() + text.length
  });

  // apply the updated entity to the content
  return Modifier.applyEntity(
    updatedContentState,
    updatedEntitySelection,
    entityKey
  );
};

export default replaceText;
