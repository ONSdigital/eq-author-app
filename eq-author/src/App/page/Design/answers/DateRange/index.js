import React from "react";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import gql from "graphql-tag";
import styled from "styled-components";
import { useMutation } from "@apollo/react-hooks";

import Date from "../Date";
import UPDATE_ANSWER from "graphql/updateAnswer.graphql";
import EarliestDateValidationRule from "graphql/fragments/earliest-date-validation-rule.graphql";
import LatestDateValidationRule from "graphql/fragments/latest-date-validation-rule.graphql";
import MinDurationValidationRule from "graphql/fragments/min-duration-validation-rule.graphql";
import MaxDurationValidationRule from "graphql/fragments/max-duration-validation-rule.graphql";
import AnswerProperties from "components/AnswerContent/AnswerProperties";
import AdvancedProperties from "components/AnswerContent/AdvancedProperties";
import AnswerValidation from "App/page/Design/Validation/AnswerValidation";
import Fallback from "components/AnswerContent/FallbackProperty/Fallback";

const Wrapper = styled.div`
  width: 100%;
`;

const DateRange = ({ answer, metadata, ...otherProps }) => {
  const [updateAnswer] = useMutation(UPDATE_ANSWER);

  return (
    <Wrapper data-test="date-range-editor">
      <Date
        key={`from-${answer.id}`}
        answer={answer}
        name="label"
        showDay
        showMonth
        showYear
        {...otherProps}
      />
      <Date
        key={`to-${answer.id}`}
        answer={answer}
        name="secondaryLabel"
        showDay
        showMonth
        showYear
        {...otherProps}
      />
      <AnswerProperties answer={answer} updateAnswer={updateAnswer} />
      <AdvancedProperties answer={answer} updateAnswer={updateAnswer}>
        <AnswerValidation answer={answer} />
        <Fallback
          answer={answer}
          metadata={metadata}
          updateAnswer={updateAnswer}
        />
      </AdvancedProperties>
    </Wrapper>
  );
};

DateRange.propTypes = {
  answer: CustomPropTypes.answer.isRequired,
  onUpdate: PropTypes.func.isRequired,
  metadata: PropTypes.array, //eslint-disable-line
};

DateRange.fragments = {
  DateRange: gql`
    fragment DateRange on Answer {
      id
      ... on BasicAnswer {
        validation {
          ... on DateRangeValidation {
            earliestDate {
              enabled
              ...EarliestDateValidationRule
            }
            latestDate {
              enabled
              ...LatestDateValidationRule
            }
            minDuration {
              enabled
              ...MinDurationValidationRule
            }
            maxDuration {
              enabled
              ...MaxDurationValidationRule
            }
          }
        }
      }
    }
    ${Date.fragments.Date}
    ${EarliestDateValidationRule}
    ${LatestDateValidationRule}
    ${MinDurationValidationRule}
    ${MaxDurationValidationRule}
  `,
};

export default DateRange;
