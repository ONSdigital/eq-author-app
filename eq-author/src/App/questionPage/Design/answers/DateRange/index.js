import React from "react";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import gql from "graphql-tag";
import styled from "styled-components";

import Date from "App/questionPage/Design/answers/Date";

import EarliestDateValidationRule from "graphql/fragments/earliest-date-validation-rule.graphql";
import LatestDateValidationRule from "graphql/fragments/latest-date-validation-rule.graphql";
import MinDurationValidationRule from "graphql/fragments/min-duration-validation-rule.graphql";
import MaxDurationValidationRule from "graphql/fragments/max-duration-validation-rule.graphql";

const Wrapper = styled.div`
  width: 100%;
`;

const DateRange = ({ answer, ...otherProps }) => (
  <Wrapper data-test="date-range-editor">
    <Date
      key={`from-${answer.id}`}
      answer={answer}
      name="label"
      showDay
      showMonth
      showYear
      {...otherProps}
    />
    <Date
      key={`to-${answer.id}`}
      answer={answer}
      name="secondaryLabel"
      showDay
      showMonth
      showYear
      {...otherProps}
    />
  </Wrapper>
);

DateRange.propTypes = {
  answer: CustomPropTypes.answer.isRequired,
  onUpdate: PropTypes.func.isRequired,
};

DateRange.fragments = {
  DateRange: gql`
    fragment DateRange on Answer {
      id
      ... on BasicAnswer {
        validation {
          ... on DateRangeValidation {
            earliestDate {
              enabled
              ...EarliestDateValidationRule
            }
            latestDate {
              enabled
              ...LatestDateValidationRule
            }
            minDuration {
              enabled
              ...MinDurationValidationRule
            }
            maxDuration {
              enabled
              ...MaxDurationValidationRule
            }
          }
        }
      }
    }
    ${Date.fragments.Date}
    ${EarliestDateValidationRule}
    ${LatestDateValidationRule}
    ${MinDurationValidationRule}
    ${MaxDurationValidationRule}
  `,
};

export default DateRange;
