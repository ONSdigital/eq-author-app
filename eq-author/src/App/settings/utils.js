export const getThemeSettingsErrorCount = (questionnaire) =>
  questionnaire?.themeSettings?.validationErrorInfo?.totalCount +
    questionnaire?.validationErrorInfo?.totalCount || 0;
