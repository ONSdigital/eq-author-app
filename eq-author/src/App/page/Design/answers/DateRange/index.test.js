import React from "react";
import { shallow } from "enzyme";
import DateRange from "./";
import { DATE_LABEL_REQUIRED } from "constants/validationMessages";
import { render as rtlRender } from "tests/utils/rtl";

const mockUseMutation = jest.fn();

jest.mock("@apollo/react-hooks", () => ({
  useMutation: () => [mockUseMutation],
}));

describe("DateRange", () => {
  let handleChange, handleUpdate, props;

  beforeEach(() => {
    handleChange = jest.fn();
    handleUpdate = jest.fn();

    props = {
      answer: {
        id: "1",
        label: "",
        type: "DateRange",
        description: "test",
        guidance: "",
        secondaryLabel: "",
        secondaryLabelDefault: "",
        properties: { required: false },
        displayName: "Untitled answer",
        advancedProperties: true,
        qCode: "yuky",
        validationErrorInfo: {
          errors: [],
        },
      },
      metadata: [],
    };
  });

  it("should render", () => {
    const wrapper = shallow(
      <DateRange
        onChange={handleChange}
        onUpdate={handleUpdate}
        answer={props.answer}
        metadata={props.metadata}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should render an error message when date label input is empty", () => {
    const error = [
      {
        errorCode: "ERR_VALID_REQUIRED",
        field: "label",
        id: "71b5f8a2-5de1-4a9a-8923-acd7fd3b7080",
        type: "answer",
        __typename: "ValidationError",
      },
    ];
    props.answer.validationErrorInfo.errors = error;

    const { getByText } = rtlRender(
      <DateRange onUpdate={handleUpdate} {...props} />
    );

    expect(getByText(DATE_LABEL_REQUIRED)).toBeTruthy();
  });

  it("should render an error message when date secondarylabel input is empty", () => {
    const error = [
      {
        errorCode: "ERR_VALID_REQUIRED",
        field: "secondaryLabel",
        id: "71b5f8a2-5de1-4a9a-8923-acd7fd3b7081",
        type: "answer",
        __typename: "ValidationError",
      },
    ];
    props.answer.validationErrorInfo.errors = error;

    const { getByText } = rtlRender(
      <DateRange onUpdate={handleUpdate} {...props} />
    );

    expect(getByText(DATE_LABEL_REQUIRED)).toBeTruthy();
  });

  it("Does not have a mutually exclusive option toggle", () => {
    const { queryByTestId } = rtlRender(
      <DateRange onUpdate={handleUpdate} {...props} />
    );

    expect(queryByTestId("toggle-or-option-date")).not.toBeInTheDocument();
  });
});
