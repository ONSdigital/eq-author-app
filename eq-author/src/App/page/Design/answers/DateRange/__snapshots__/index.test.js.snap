// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRange should render 1`] = `
<DateRange__Wrapper
  data-test="date-range-editor"
>
  <Connect(withEntityEditor(withValidationError(UnwrappedDate)))
    answer={
      Object {
        "advancedProperties": true,
        "description": "test",
        "displayName": "Untitled answer",
        "guidance": "",
        "id": "1",
        "label": "",
        "properties": Object {
          "required": false,
        },
        "qCode": "yuky",
        "secondaryLabel": "",
        "secondaryLabelDefault": "",
        "type": "DateRange",
        "validationErrorInfo": Object {
          "errors": Array [],
        },
      }
    }
    key="from-1"
    name="label"
    onChange={[MockFunction]}
    onUpdate={[MockFunction]}
    showDay={true}
    showMonth={true}
    showYear={true}
  />
  <Connect(withEntityEditor(withValidationError(UnwrappedDate)))
    answer={
      Object {
        "advancedProperties": true,
        "description": "test",
        "displayName": "Untitled answer",
        "guidance": "",
        "id": "1",
        "label": "",
        "properties": Object {
          "required": false,
        },
        "qCode": "yuky",
        "secondaryLabel": "",
        "secondaryLabelDefault": "",
        "type": "DateRange",
        "validationErrorInfo": Object {
          "errors": Array [],
        },
      }
    }
    key="to-1"
    name="secondaryLabel"
    onChange={[MockFunction]}
    onUpdate={[MockFunction]}
    showDay={true}
    showMonth={true}
    showYear={true}
  />
  <AnswerProperties
    answer={
      Object {
        "advancedProperties": true,
        "description": "test",
        "displayName": "Untitled answer",
        "guidance": "",
        "id": "1",
        "label": "",
        "properties": Object {
          "required": false,
        },
        "qCode": "yuky",
        "secondaryLabel": "",
        "secondaryLabelDefault": "",
        "type": "DateRange",
        "validationErrorInfo": Object {
          "errors": Array [],
        },
      }
    }
    updateAnswer={[MockFunction]}
  />
  <AdvancedProperties
    answer={
      Object {
        "advancedProperties": true,
        "description": "test",
        "displayName": "Untitled answer",
        "guidance": "",
        "id": "1",
        "label": "",
        "properties": Object {
          "required": false,
        },
        "qCode": "yuky",
        "secondaryLabel": "",
        "secondaryLabelDefault": "",
        "type": "DateRange",
        "validationErrorInfo": Object {
          "errors": Array [],
        },
      }
    }
    enableHorizontalRule={true}
    updateAnswer={[MockFunction]}
  >
    <Memo(AnswerValidation)
      answer={
        Object {
          "advancedProperties": true,
          "description": "test",
          "displayName": "Untitled answer",
          "guidance": "",
          "id": "1",
          "label": "",
          "properties": Object {
            "required": false,
          },
          "qCode": "yuky",
          "secondaryLabel": "",
          "secondaryLabelDefault": "",
          "type": "DateRange",
          "validationErrorInfo": Object {
            "errors": Array [],
          },
        }
      }
    />
    <Fallback
      answer={
        Object {
          "advancedProperties": true,
          "description": "test",
          "displayName": "Untitled answer",
          "guidance": "",
          "id": "1",
          "label": "",
          "properties": Object {
            "required": false,
          },
          "qCode": "yuky",
          "secondaryLabel": "",
          "secondaryLabelDefault": "",
          "type": "DateRange",
          "validationErrorInfo": Object {
            "errors": Array [],
          },
        }
      }
      metadata={Array []}
      updateAnswer={[MockFunction]}
    />
  </AdvancedProperties>
</DateRange__Wrapper>
`;
